icu4c/source/common/uresbund.cpp

```diff
@@ -91,6 +91,173 @@ static UBool chopLocale(char *name) {
     return FALSE;
 }
 
+static UBool hasScript(const char* localeID) {
+    UErrorCode err = U_ZERO_ERROR;
+    int32_t scriptLength = uloc_getScript(localeID, NULL, 0, &err);
+    return scriptLength != 0;
+}
+
+static UBool hasRegion(const char* localeID) {
+    UErrorCode err = U_ZERO_ERROR;
+    int32_t regionLength = uloc_getCountry(localeID, NULL, 0, &err);
+    return regionLength != 0;
+}
+
+static UBool hasVariant(const char* localeID) {
+    UErrorCode err = U_ZERO_ERROR;
+    int32_t variantLength = uloc_getVariant(localeID, NULL, 0, &err);
+    return variantLength != 0;
+}
+
+// This file contains the tables for doing locale fallback, which are generated
+// by the CLDR-to-ICU process directly from the CLDR data.  This file should only
+// ever be included from here.
+#include "localeFallbackTables.h"
+
+static const char* performFallbackLookup(const char* key,
+                                         const char* keyStrs,
+                                         const char* valueStrs,
+                                         const int32_t* lookupTable,
+                                         int32_t lookupTableSize) {
+    const int32_t* bottom = lookupTable;
+    const int32_t* top = lookupTable + lookupTableSize;
+
+    while (bottom < top) {
+        const int32_t* middle = bottom + ((top - bottom) / 2);
+        middle -= (middle - lookupTable) % 2;
```

Comment:
>You could also write it directly with string literals, instead of the index indirection; I would hope that with enough const keywords the compiler will de-duplicate read-only string literals.

After I wrote this, I thought a lot about that too-- am I really gaining anything by doing all the string management myself-- wouldn't I get basically the same thing as what I'm getting right now by jumping through all these hoops?

Your comment and this thought of mine set off a giant wild goose chase that ate up most of my day and left me tearing my hair out.

Just getting rid of the separate `char*` arrays and making the main tables into arrays of `char*` looked good and worked fine, but blew up the size of the generated file by 13K, from 53K to 66K.  I'm not sure why.  I don't think it's because the compiler isn't deduping all those string literals-- I checked that in a small test app.  I think part of it may by that the main tables went from being arrays of 32-bit integers to arrays of 64-bit pointers, but I don't think those tables are big enough for a doubling in their size to add 13K.

I also couldn't figure out a good way to make the table into an array of two-entry arrays.  C++ apparently doesn't like this, or I simply couldn't figure out the declarator syntax.  I tried making the whole thing into an array of `std::pair`, but this gave me a "global constructor" warning-- according to StackOverflow, this is a warning that I'm introducing static-init code I didn't need before.  I was really trying to avoid this kind of thing, so I gave up, stuck with what I had, and just made the change you suggest above.

If somebody smarter than I wants to come up with some better here, great, but I'm reluctant at this point to throw good time after bad.

Reply:
Sounds like that is best left for a future possible optimization.